### PR TITLE
Allow linking against system libraries for almost everything

### PR DESCRIPTION
**Pull request by Tatsh**
This allows linking to the system libraries (even if they may not work) for almost everything except Lua.

Just to note: libpng 1.6 definitely does not work for the time being (crashes on launch).

The code is updated to support ffmpeg 4.1+, and retains backward compatibility.

All CMake files have been formatted with [cmake-format](https://github.com/cheshirekow/cmake_format).

There are also some minor fixes which affect Linux only.

Tested with the following on Linux:

```
cmake -DCMAKE_BUILD_TYPE=Release -DWITH_FFMPEG=yes -DWITH_FFMPEG_JOBS=13

cmake -DCMAKE_BUILD_TYPE=Release -DWITH_FFMPEG=no

cmake -G Ninja -DCMAKE_BUILD_TYPE=RelWithDebugInfo -DWITH_GTK2=yes -DWITH_GPL_LIBS=yes -DWITH_WAV=yes -DWITH_MP3=yes -DWITH_OGG=yes -DWITH_SDL=yes -DWITH_FFMPEG=yes -DWITH_FFMPEG_JOBS=13 -DCMAKE_VERBOSE_MAKEFILE=yes -DWITH_SYSTEM_GLEW=yes -DWITH_SYSTEM_TOMMATH=yes -DWITH_SYSTEM_MAD=yes -DWITH_SYSTEM_ZLIB=yes -DWITH_SYSTEM_JSONCPP=no -DWITH_SYSTEM_PNG=no -DWITH_SYSTEM_JPEG=yes -DWITH_SYSTEM_PCRE=yes -DWITH_SYSTEM_OGG=yes -DWITH_SYSTEM_FFMPEG=yes -DWITH_FULL_RELEASE=yes -DWITH_CRASH_HANDLER=yes -DWITH_NETWORKING=yes -DWITH_SSE2=yes -DWITH_XINERAMA=yes -DWITH_TTY=no -DWITH_GLES2=yes -DWITH_MINIMAID=yes -DWITH_SDL=yes -DWITH_LTO=yes ..
```

Yes, LTO works.

### CODE REVIEW

**Comment by Tatsh** [comment]
I propose complete removal of autohell code. I am happy to make that change.

**Comment by shakesoda** [comment]
this is a concerning amount of things to merge at once but i'm pressing the button anyways.

thanks 👍

**Comment by sl1pkn07** [comment]
~~~
[  7%] Linking CXX executable /tmp/makepkg/sl1-stepmania-git/src/stepmania/stepmania
/usr/bin/ld: CMakeFiles/StepMania.dir/NotesWriterJson.cpp.o: in function `Serialize(BackgroundChange const&, Json::Value&)':
NotesWriterJson.cpp:(.text+0x5a): undefined reference to `Json::Value::operator=(Json::Value const&)'
/usr/bin/ld: NotesWriterJson.cpp:(.text+0x8c): undefined reference to `Json::Value::operator=(Json::Value const&)'
/usr/bin/ld: NotesWriterJson.cpp:(.text+0xbe): undefined reference to `Json::Value::operator=(Json::Value const&)'
/usr/bin/ld: NotesWriterJson.cpp:(.text+0xf0): undefined reference to `Json::Value::operator=(Json::Value const&)'
/usr/bin/ld: NotesWriterJson.cpp:(.text+0x12a): undefined reference to `Json::Value::operator=(Json::Value const&)'
/usr/bin/ld: CMakeFiles/StepMania.dir/NotesWriterJson.cpp.o:NotesWriterJson.cpp:(.text+0x164): more undefined references to `Json::Value::operator=(Json::Value const&)' follow
collect2: error: ld returned 1 exit status
make[2]: *** [src/CMakeFiles/StepMania.dir/build.make:6369: /tmp/makepkg/sl1-stepmania-git/src/stepmania/stepmania] Error 1
make[1]: *** [CMakeFiles/Makefile2:146: src/CMakeFiles/StepMania.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
~~~

with WITH_SYSTEM_JSONCPP=ON and system json 1.8.4

**Comment by Tatsh** [comment]
SYSTEM_JSON with any version is not supported at this time. The version of jsoncpp that comes with StepMania must be used.

**Comment by sl1pkn07** [comment]
ok. build it ok, but fails in run (after make install)

~~~
┌─┤[$]|[sl1pkn07]|[sL1pKn07]|[/usr/share/stepmania-5.1]|
└───╼  stepmania
StepMania5.1-git-c0c26697a1
Compiled 20190314 @ 14:43:42 (build c0c26697a1)
Log starting 2019-03-14 15:49:03
OS: Linux ver 050000
Crash backtrace component: x86 custom backtrace
Crash lookup component: backtrace_symbols
Runtime library: glibc 2.28
Threads library: NPTL 2.28
libavcodec: 0x3a2364 (3810148)
/////////////////////////////////////////
WARNING: Selected theme '_fallback' not found.  Trying Theme preference default value '_fallback'.
/////////////////////////////////////////
/////////////////////////////////////////
WARNING: Theme preference defaults to '_fallback', which cannot be used.  Trying DefaultTheme preference 'default'.
/////////////////////////////////////////
/////////////////////////////////////////
WARNING: ReadFile(Save/ThemePrefs.ini): No such file or directory
/////////////////////////////////////////
Fatal: crash from within the crash signal handler
~~~
coredumpctlr say nothing about stepmania segfaults

my build options:

~~~~
  cmake ../stepmania \
    -DCMAKE_BUILD_TYPE=None \
    -DCMAKE_INSTALL_PREFIX=/usr/share \
    -DCMAKE_SKIP_RPATH=OFF \
    -DOpenGL_GL_PREFERENCE=GLVND \
    -DWITH_PULSEAUDIO=ON \
    -DWITH_GTK2=OFF \
    -DWITH_SDL=ON \
    -DWITH_JACK=ON \
    -DWITH_LTO=ON \
    -DWITH_SYSTEM_FFMPEG=ON \
    -DWITH_SYSTEM_GLEW=ON \
    -DWITH_SYSTEM_JPEG=ON \
    -DWITH_SYSTEM_JSONCPP=OFF \
    -DWITH_SYSTEM_MAD=ON \
    -DWITH_SYSTEM_OGG=ON \
    -DWITH_SYSTEM_PCRE=ON \
    -DWITH_SYSTEM_PNG=ON \
    -DWITH_SYSTEM_TOMCRYPT=ON \
    -DWITH_SYSTEM_TOMMATH=ON \
    -DWITH_SYSTEM_ZLIB=ON

**Comment by Tatsh** [comment]
Build without the crash handler, and with -DCMAKE_BUILD_TYPE=RelWithDebInfo

Use gdb to get a backtrace:

gdb stepmania
r
(Wait for crash)
bt

Paste the output here.

**Comment by Tatsh** [comment]
Actually I think I know what it is. You probably have libpng 1.6. You can either install libpng 1.5 alongside, downgrade your system to libpng 1.5 if that is possible, or build with the included libpng.

libpng 1.6+ is not supported. It's not being worked on because eventually all image loading will be done with stb_image.

**Comment by sl1pkn07** [comment]
yep. finaly i manage to run stepmania with bundle libpng. yes, i have installed libpng 1.6.36

testing now with libpng 1.5.28

greetings

File: StepmaniaCore.cmake [hunk 1]
```diff
@@ -1,463 +1,460 @@
...
-# Allow for finding our libraries in a standard location.
-list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}"
-            "${SM_CMAKE_DIR}/Modules/")
-
-include("${SM_CMAKE_DIR}/DefineOptions.cmake")
-
-include("${SM_CMAKE_DIR}/SMDefs.cmake")
-
-# Put the predefined targets in separate groups.
-set_property(GLOBAL PROPERTY USE_FOLDERS ON)
```
Comment (Prcuvu) [inline StepmaniaCore.cmake]: Why did you choose to disable USE_FOLDERS?

**Pull request merged**